bft_types/src/lib.rs

```diff
@@ -97,53 +136,121 @@ impl Display for LocalisedInstruction {
+    name: PathBuf,
+    /// A vector of instructions. Not sure how else to describe it
+    instructions: Vec<LocalisedInstruction>,
+    /// Hashmap to record the index of jumps in instructions, and the indexes of their counterparts
+    jump_map: HashMap<usize, usize>,
 }
 
 impl BfProgram {
-    // TODO: check file_path typing
-    /// Attempt to load a valid Brainfuck program from the specified file path
-    pub fn from_file<P: AsRef<Path>>(file_path: P) -> std::io::Result<BfProgram> {
-        let file_contents = fs::read_to_string(&file_path)?;
-        Ok(Self::new(file_path, file_contents.as_str()))
+    /// Attempt to load a valid Brainfuck program from the specified file path.
+    ///
+    /// ```no_run
+    ///# use bft_types::BfProgram;
+    ///# use std::path::PathBuf;
+    ///#  
+    ///  let bf_file = PathBuf::from("my_bf_program.bf");
+    ///
+    ///  let my_bf_program = BfProgram::from_file(bf_file);
+    /// ```
+    pub fn from_file<P: AsRef<Path>>(file_path: P) -> std::io::Result<BfProgram> {
+        let file_contents = fs::read_to_string(&file_path)?;
+        Ok(Self::new(file_path, file_contents.as_str()))
     }
 
     /// Construct a new BfProgram from a &str
-    fn new<P: AsRef<Path>>(filename: P, file_contents: &str) -> Self {
-        let mut instructions: Vec<LocalisedInstruction> = Vec::new();
+    fn new<P: AsRef<Path>>(filename: P, file_contents: &str) -> Self {
+        let mut instructions: Vec<LocalisedInstruction> = Vec::new();
+        let jump_map = HashMap::<usize, usize>::new();
 
         // TODO: see if this can be shortened
         for (line_number, file_line) in file_contents.lines().enumerate() {
             for (col_number, character) in file_line.chars().enumerate() {
-                match Instruction::from_char(character) {
+                match Instruction::from_char(character) {
                     None => (),
                     Some(instr) => {
-                        instructions.push(LocalisedInstruction::new(
-                            instr,
-                            line_number,
-                            col_number,
-                        ));
+                        instructions.push(LocalisedInstruction::new(
+                            instr,
+                            line_number + 1,
+                            col_number + 1,
+                        ));
                     }
                 }
             }
         }
 
         Self {
-            name: filename.as_ref().to_path_buf(),
+            name: filename.as_ref().to_path_buf(),
             instructions,
+            jump_map,
         }
     }
 
     /// Get the name of the program
-    pub fn name(&self) -> &Path {
+    pub fn name(&self) -> &Path {
         &self.name
     }
 
-    /// Borrow(?) a copy of the instructions
-    pub fn instructions(&self) -> &[LocalisedInstruction] {
-        &self.instructions
+    /// Borrow a copy of the instructions
```

Comment:
It's pretty clear this is a borrow (the return type tells us)

```suggestion
    /// The instructions which make up this program
```